providers/fips/self_test_kats.c

```diff
@@ -647,6 +647,204 @@ static int self_test_asym_cipher(const ST_KAT_ASYM_CIPHER *t, OSSL_SELF_TEST *st
     return ret;
 }
 
+/*
+ * FIPS 140-3 IG 10.3.A resolution 14 mandates a CAST for ML-KEM
+ * encapsulation.
+ */
+static int self_test_kem_encapsulate(const ST_KAT_KEM *t, OSSL_SELF_TEST *st,
+                                     OSSL_LIB_CTX *libctx, EVP_PKEY *pkey)
+{
+    int ret = 0;
+    EVP_PKEY_CTX *ctx;
+    unsigned char *wrapped = NULL, *secret = NULL;
+    size_t wrappedlen, secretlen;
+    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+
+    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_KAT_KEM, "KEM_Encap");
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, "");
+    if (ctx == NULL)
+        goto err;
+
+    *params = OSSL_PARAM_construct_octet_string(OSSL_KEM_PARAM_IKME,
+                                                (unsigned char *)t->entropy,
+                                                t->entropy_len);
+    if (EVP_PKEY_encapsulate_init(ctx, params) <= 0)
+        goto err;
+
+    /* Get output sizes */
+    if (EVP_PKEY_encapsulate(ctx, NULL, &wrappedlen, NULL, &secretlen) <= 0)
+        goto err;
+    wrapped = OPENSSL_malloc(wrappedlen);
+    secret = OPENSSL_malloc(secretlen);
+    if (wrapped == NULL || secret == NULL)
+        goto err;
+
+    /* Encapsulate */
+    if (EVP_PKEY_encapsulate(ctx, wrapped, &wrappedlen, secret, &secretlen) <= 0)
+        goto err;
+
+    /* Compare outputs */
+    OSSL_SELF_TEST_oncorrupt_byte(st, wrapped);
+    if (wrappedlen != t->cipher_text_len
+            || memcmp(wrapped, t->cipher_text, t->cipher_text_len) != 0)
+        goto err;
+
+    OSSL_SELF_TEST_oncorrupt_byte(st, secret);
+    if (secretlen != t->secret_len
+            || memcmp(secret, t->secret, t->secret_len) != 0)
+        goto err;
+
+    ret = 1;
+ err:
+    OPENSSL_free(wrapped);
+    OPENSSL_free(secret);
+    EVP_PKEY_CTX_free(ctx);
+    OSSL_SELF_TEST_onend(st, ret);
+    return ret;
+}
+
+/*
+ * FIPS 140-3 IG 10.3.A resolution 14 mandates a CAST for ML-KEM
+ * decapsulation both for the rejection path and the normal path.
+ */
+static int self_test_kem_decapsulate(const ST_KAT_KEM *t, OSSL_SELF_TEST *st,
+                                     OSSL_LIB_CTX *libctx, EVP_PKEY *pkey,
+                                     int reject)
```

Comment:
pass might be a better name? (and invert the logic),

Reply:
NIST uses the term _reject_ for the failure path.  They don't name the success path, so I think it is better staying as reject.

Reply:
Do you want to rename it now that it is better understood?